File: server/src/controllers/trade.controller.js
import { asyncHandler } from '../utils/async-handler.js';
import { ApiResponse } from '../utils/api-response.js';
import { ApiError } from '../utils/api-error.js';
import { userModel } from '../models/user.models.js';
import { portfolioModel } from '../models/portfolio.model.js';
import { transactionModel } from '../models/transaction.model.js';

/**
 * POST /api/v1/trade/buy
 * Body: { symbol, quantity, price }
 */
export const buyStock = asyncHandler(async (req, res) => {
  const { symbol, quantity, price } = req.body;
  const userId = req.user.id;

  // Validate inputs
  if (!symbol || !quantity || !price) {
    return res.status(400).json(new ApiError(400, 'Symbol, quantity, and price are required'));
  }

  const qty = parseInt(quantity);
  const unitPrice = parseFloat(price);

  if (qty < 1) {
    return res.status(400).json(new ApiError(400, 'Quantity must be at least 1'));
  }
  if (unitPrice <= 0) {
    return res.status(400).json(new ApiError(400, 'Price must be greater than 0'));
  }

  const totalCost = qty * unitPrice;

  // Get user and check balance
  const user = await userModel.findById(userId);
  if (!user) {
    return res.status(404).json(new ApiError(404, 'User not found'));
  }

  if (user.balance < totalCost) {
    return res.status(400).json(
      new ApiError(400, `Insufficient balance. You have $${user.balance.toFixed(2)} but need $${totalCost.toFixed(2)}`)
    );
  }

  // Deduct balance
  user.balance -= totalCost;
  await user.save();

  // Update or create portfolio entry
  let portfolio = await portfolioModel.findOne({ user: userId, symbol: symbol.toUpperCase() });

  if (portfolio) {
    // Recalculate average buy price
    const newTotalInvested = portfolio.totalInvested + totalCost;
    const newQuantity = portfolio.quantity + qty;
    portfolio.avgBuyPrice = newTotalInvested / newQuantity;
    portfolio.quantity = newQuantity;
    portfolio.totalInvested = newTotalInvested;
    await portfolio.save();
  } else {
    portfolio = await portfolioModel.create({
      user: userId,
      symbol: symbol.toUpperCase(),
      quantity: qty,
      avgBuyPrice: unitPrice,
      totalInvested: totalCost,
    });
  }

  // Record transaction
  const transaction = await transactionModel.create({
    user: userId,
    symbol: symbol.toUpperCase(),
    type: 'buy',
    quantity: qty,
    price: unitPrice,
    total: totalCost,
    balanceAfter: user.balance,
  });

  return res.status(200).json(
    new ApiResponse(200, {
      transaction,
      portfolio,
      balance: user.balance,
    }, `Successfully bought ${qty} share(s) of ${symbol.toUpperCase()}`)
  );
});

/**
 * POST /api/v1/trade/sell
 * Body: { symbol, quantity, price }
 */
export const sellStock = asyncHandler(async (req, res) => {
  const { symbol, quantity, price } = req.body;
  const userId = req.user.id;

  // Validate inputs
  if (!symbol || !quantity || !price) {
    return res.status(400).json(new ApiError(400, 'Symbol, quantity, and price are required'));
  }

  const qty = parseInt(quantity);
  const unitPrice = parseFloat(price);

  if (qty < 1) {
    return res.status(400).json(new ApiError(400, 'Quantity must be at least 1'));
  }
  if (unitPrice <= 0) {
    return res.status(400).json(new ApiError(400, 'Price must be greater than 0'));
  }

  // Check portfolio holdings
  const portfolio = await portfolioModel.findOne({ user: userId, symbol: symbol.toUpperCase() });

  if (!portfolio || portfolio.quantity === 0) {
    return res.status(400).json(
      new ApiError(400, `You don't own any shares of ${symbol.toUpperCase()}`)
    );
  }

  if (portfolio.quantity < qty) {
    return res.status(400).json(
      new ApiError(400, `Insufficient shares. You own ${portfolio.quantity} share(s) of ${symbol.toUpperCase()} but tried to sell ${qty}`)
    );
  }

  const totalRevenue = qty * unitPrice;

  // Credit balance
  const user = await userModel.findById(userId);
  if (!user) {
    return res.status(404).json(new ApiError(404, 'User not found'));
  }

  user.balance += totalRevenue;
  await user.save();

  // Update portfolio
  portfolio.quantity -= qty;
  portfolio.totalInvested -= qty * portfolio.avgBuyPrice;

  if (portfolio.quantity === 0) {
    portfolio.avgBuyPrice = 0;
    portfolio.totalInvested = 0;
  }

  await portfolio.save();

  // Record transaction
  const transaction = await transactionModel.create({
    user: userId,
    symbol: symbol.toUpperCase(),
    type: 'sell',
    quantity: qty,
    price: unitPrice,
    total: totalRevenue,
    balanceAfter: user.balance,
  });

  return res.status(200).json(
    new ApiResponse(200, {
      transaction,
      portfolio,
      balance: user.balance,
    }, `Successfully sold ${qty} share(s) of ${symbol.toUpperCase()}`)
  );
});

/**
 * GET /api/v1/trade/portfolio
 * Returns all holdings for the logged-in user
 */
export const getPortfolio = asyncHandler(async (req, res) => {
  const userId = req.user.id;

  const holdings = await portfolioModel.find({ user: userId, quantity: { $gt: 0 } }).sort({ updatedAt: -1 });

  return res.status(200).json(
    new ApiResponse(200, holdings, 'Portfolio fetched successfully')
  );
});

/**
 * GET /api/v1/trade/portfolio/:symbol
 * Returns holding for a specific symbol
 */
export const getHolding = asyncHandler(async (req, res) => {
  const userId = req.user.id;
  const { symbol } = req.params;

  const holding = await portfolioModel.findOne({
    user: userId,
    symbol: symbol.toUpperCase(),
  });

  return res.status(200).json(
    new ApiResponse(200, holding || { quantity: 0, avgBuyPrice: 0, totalInvested: 0 }, 'Holding fetched successfully')
  );
});

/**
 * GET /api/v1/trade/transactions
 * Returns all transaction history for the logged-in user
 */
export const getTransactions = asyncHandler(async (req, res) => {
  const userId = req.user.id;

  const transactions = await transactionModel.find({ user: userId }).sort({ createdAt: -1 });

  return res.status(200).json(
    new ApiResponse(200, transactions, 'Transactions fetched successfully')
  );
});

File: corebackend/Controllers/TradeController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using MongoDB.Driver;
using System.Security.Claims;
using corebackend.Models;
using corebackend.Models.DTOs;
using corebackend.Services;

namespace corebackend.Controllers
{
    [ApiController]
    [Route("api/v1/trade")]
    [Authorize]
    public class TradeController : ControllerBase
    {
        private readonly IMongoCollection<User> _users;
        private readonly IMongoCollection<Portfolio> _portfolios;
        private readonly IMongoCollection<Transaction> _transactions;

        public TradeController(MongoService mongoService)
        {
            _users = mongoService.GetCollection<User>("users");
            _portfolios = mongoService.GetCollection<Portfolio>("portfolios");
            _transactions = mongoService.GetCollection<Transaction>("transactions");
        }

        private string UserId => User.FindFirstValue("id") ?? "";

        // POST /api/v1/trade/buy
        [HttpPost("buy")]
        public async Task<IActionResult> BuyStock([FromBody] TradeRequest request)
        {
            var symbol = request.Symbol.ToUpper().Trim();
            var totalCost = request.Quantity * request.Price;

            var user = await _users.Find(u => u.Id == UserId).FirstOrDefaultAsync();
            if (user == null)
            {
                return StatusCode(404, new ApiErrorResponse(404, "User not found"));
            }

            if (user.Balance < totalCost)
            {
                return StatusCode(400, new ApiErrorResponse(400, $"Insufficient balance. You have ${user.Balance:F2} but need ${totalCost:F2}"));
            }

            // Deduct balance
            user.Balance -= totalCost;
            await _users.ReplaceOneAsync(u => u.Id == user.Id, user);

            // Update or create portfolio entry
            var portfolio = await _portfolios.Find(p => p.User == UserId && p.Symbol == symbol).FirstOrDefaultAsync();

            if (portfolio != null)
            {
                var newTotalInvested = portfolio.TotalInvested + totalCost;
                var newQuantity = portfolio.Quantity + request.Quantity;
                portfolio.AvgBuyPrice = newTotalInvested / newQuantity;
                portfolio.Quantity = newQuantity;
                portfolio.TotalInvested = newTotalInvested;
                portfolio.UpdatedAt = DateTime.UtcNow;
                await _portfolios.ReplaceOneAsync(p => p.Id == portfolio.Id, portfolio);
            }
            else
            {
                portfolio = new Portfolio
                {
                    User = UserId,
                    Symbol = symbol,
                    Quantity = request.Quantity,
                    AvgBuyPrice = request.Price,
                    TotalInvested = totalCost
                };
                await _portfolios.InsertOneAsync(portfolio);
            }

            // Record transaction
            var transaction = new Transaction
            {
                User = UserId,
                Symbol = symbol,
                Type = "buy",
                Quantity = request.Quantity,
                Price = request.Price,
                Total = totalCost,
                BalanceAfter = user.Balance
            };
            await _transactions.InsertOneAsync(transaction);

            return Ok(new ApiResponse<object>(200, new
            {
                transaction,
                portfolio,
                balance = user.Balance
            }, $"Successfully bought {request.Quantity} share(s) of {symbol}"));
        }

        // POST /api/v1/trade/sell
        [HttpPost("sell")]
        public async Task<IActionResult> SellStock([FromBody] TradeRequest request)
        {
            var symbol = request.Symbol.ToUpper().Trim();

            var portfolio = await _portfolios.Find(p => p.User == UserId && p.Symbol == symbol).FirstOrDefaultAsync();

            if (portfolio == null || portfolio.Quantity == 0)
            {
                return StatusCode(400, new ApiErrorResponse(400, $"You don't own any shares of {symbol}"));
            }

            if (portfolio.Quantity < request.Quantity)
            {
                return StatusCode(400, new ApiErrorResponse(400, $"Insufficient shares. You own {portfolio.Quantity} share(s) of {symbol} but tried to sell {request.Quantity}"));
            }

            var totalRevenue = request.Quantity * request.Price;

            var user = await _users.Find(u => u.Id == UserId).FirstOrDefaultAsync();
            if (user == null)
            {
                return StatusCode(404, new ApiErrorResponse(404, "User not found"));
            }

            // Credit balance
            user.Balance += totalRevenue;
            await _users.ReplaceOneAsync(u => u.Id == user.Id, user);

            // Update portfolio
            portfolio.Quantity -= request.Quantity;
            portfolio.TotalInvested -= request.Quantity * portfolio.AvgBuyPrice;

            if (portfolio.Quantity == 0)
            {
                portfolio.AvgBuyPrice = 0;
                portfolio.TotalInvested = 0;
            }

            portfolio.UpdatedAt = DateTime.UtcNow;
            await _portfolios.ReplaceOneAsync(p => p.Id == portfolio.Id, portfolio);

            // Record transaction
            var transaction = new Transaction
            {
                User = UserId,
                Symbol = symbol,
                Type = "sell",
                Quantity = request.Quantity,
                Price = request.Price,
                Total = totalRevenue,
                BalanceAfter = user.Balance
            };
            await _transactions.InsertOneAsync(transaction);

            return Ok(new ApiResponse<object>(200, new
            {
                transaction,
                portfolio,
                balance = user.Balance
            }, $"Successfully sold {request.Quantity} share(s) of {symbol}"));
        }

        // GET /api/v1/trade/portfolio
        [HttpGet("portfolio")]
        public async Task<IActionResult> GetPortfolio()
        {
            var holdings = await _portfolios.Find(p => p.User == UserId && p.Quantity > 0)
                                            .SortByDescending(p => p.UpdatedAt)
                                            .ToListAsync();

            return Ok(new ApiResponse<List<Portfolio>>(200, holdings, "Portfolio fetched successfully"));
        }

        // GET /api/v1/trade/portfolio/:symbol
        [HttpGet("portfolio/{symbol}")]
        public async Task<IActionResult> GetHolding(string symbol)
        {
            var holding = await _portfolios.Find(p => p.User == UserId && p.Symbol == symbol.ToUpper().Trim()).FirstOrDefaultAsync();

            var result = holding ?? new Portfolio { Quantity = 0, AvgBuyPrice = 0, TotalInvested = 0 };

            return Ok(new ApiResponse<Portfolio>(200, result, "Holding fetched successfully"));
        }

        // GET /api/v1/trade/transactions
        [HttpGet("transactions")]
        public async Task<IActionResult> GetTransactions()
        {
            var transactions = await _transactions.Find(t => t.User == UserId)
                                                  .SortByDescending(t => t.CreatedAt)
                                                  .ToListAsync();

            return Ok(new ApiResponse<List<Transaction>>(200, transactions, "Transactions fetched successfully"));
        }
    }
}
